File: code/82-Remove Duplicates from Sorted List II/82-Remove Duplicates from Sorted List II.py
# -*- coding: utf-8 -*
"""
@author: Kai Chen
@file: 82-Remove Duplicates from Sorted List II.py
@time: 2021/3/25 9:03
@desc:

Given the head of a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list. Return the linked list sorted as well.



Example 1:


Input: head = [1,2,3,3,4,4,5]
Output: [1,2,5]
Example 2:


Input: head = [1,1,1,2,3]
Output: [2,3]


Constraints:

The number of nodes in the list is in the range [0, 300].
-100 <= Node.val <= 100
The list is guaranteed to be sorted in ascending order.

"""

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        vals = []
        dups = []
        cur = head
        while cur:
            if cur.val not in vals:
                vals.append(cur.val)
            elif cur.val in vals and cur.val not in dups:
                dups.append(cur.val)
            cur = cur.next
        cur_ = ListNode()
        cur_.next = head
        prehead = cur_
        while cur_.next:
            if cur_.next.val not in dups:
                cur_ = cur_.next
            else:
                cur_.next = cur_.next.next
        return prehead.next

File: code/82-Remove Duplicates from Sorted List II/82-Remove Duplicates from Sorted List II.cpp
/*
LeetCode 82 - Remove Duplicates from Sorted List II

解题步骤（自动整理）
1. 用快慢指针/虚拟头节点等技巧处理链表边界
2. 按题意进行遍历、反转、合并或断链等操作
3. 返回处理后的链表头节点或计算结果
*/
/*
Given the head of a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list. Return the linked list sorted as well.



Example 1:


Input: head = [1,2,3,3,4,4,5]
Output: [1,2,5]
Example 2:


Input: head = [1,1,1,2,3]
Output: [2,3]


Constraints:

The number of nodes in the list is in the range [0, 300].
-100 <= Node.val <= 100
The list is guaranteed to be sorted in ascending order.
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
      ListNode *deleteDuplicates(ListNode *head) {
          if (!head || !head->next) return head;
          ListNode *p = head->next;
          if (head->val == p->val) {
              while (p && head->val == p->val) {
                  ListNode *tmp = p;
                  p = p->next;
                  delete tmp;
              }
              delete head;
              return deleteDuplicates(p);
          } else {
              head->next = deleteDuplicates(head->next);
              return head;
            }
      }
};
